ami/models/hifigan_discriminators.py

```diff
@@ -0,0 +1,220 @@
+# Ref:
+# https://github.com/jik876/hifi-gan/blob/master/models.py
+
+import logging
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.utils.parametrizations import spectral_norm, weight_norm
+
+
+def get_padding(kernel_size: int, dilation: int = 1) -> int:
+    return int((kernel_size * dilation - dilation) / 2)
+
+
+class PeriodDiscriminator(nn.Module):
+    """PeriodDiscriminator that constitutes MultiPeriodDiscriminator described
+    below.
+
+    Discriminate authenticity of the periodicity of the input audio.
+    """
+
+    def __init__(
+        self,
+        period: int,
+        in_channels: int = 1,
+        kernel_size: int = 5,
+        stride: int = 3,
+    ) -> None:
+        super().__init__()
+        self.period = period
+
+        channels = [
+            (in_channels, 32),
+            (32, 128),
+            (128, 512),
+            (512, 1024),
+            (1024, 1024),
+        ]
+        self.convs = nn.ModuleList()
+        for i, (in_ch, out_ch) in enumerate(channels, start=1):
+            self.convs.append(
+                weight_norm(
+                    nn.Conv1d(
+                        in_ch,
+                        out_ch,
+                        kernel_size,
+                        stride=1 if i == len(channels) else stride,
+                        padding=2 if i == len(channels) else get_padding(kernel_size, 1),
+                    )
+                )
+            )
+        self.conv_post = weight_norm(nn.Conv1d(channels[-1][-1], 1, 3, 1, padding=1))
+
+    def forward(self, input_waveforms: torch.Tensor) -> tuple[torch.Tensor, list[torch.Tensor]]:
+        """Discriminate authenticity of the periodicity of the input audio.
+
+        Args:
+            input_waveforms (torch.Tensor):
+                Input waveforms. Shape: [batch_size, in_channels, sample_size]
+
+        Returns:
+            torch.Tensor:
+                Authenticities. Shape: [batch_size, frames]
+            list[torch.Tensor]:
+                Feature maps from layers. Shape of each tensor in list: [batch_size, channels, each_n_frames, self.period]
+        """
+        fmaps: list[torch.Tensor] = []
+        x = input_waveforms
+
+        # 1d to 2d
+        batch_size, channels, sample_size = x.shape
+        if sample_size % self.period != 0:  # pad first
+            n_pad = self.period - (sample_size % self.period)
+            x = F.pad(x, (0, n_pad), "reflect")
+            sample_size = sample_size + n_pad
+        x = x.view(batch_size, channels, sample_size // self.period, self.period)
+        x = x.movedim(-1, 0)
+        x = x.contiguous().view(batch_size * self.period, channels, sample_size // self.period)
+
+        for conv in self.convs:
+            x = conv(x)
+            x = F.leaky_relu(x, 0.1)
+            fmap = x.contiguous().view(self.period, batch_size, x.size(-2), x.size(-1))
```

Comment:
ここは`contiguous`いらないと思う。心配なら`view`じゃなく`reshape`にするとか
```suggestion
            fmap = x.view(self.period, batch_size, x.size(-2), x.size(-1))
```

Reply:
https://github.com/MLShukai/ami/pull/242/commits/676f67d9bcbaac4b812415342fee7106a0dd924f で修正